### ISSUE TITLE
SplatNode vs RestParameterNode

### ISSUE THREAD

**Issue by seven1m** [ruby/prism#1635]
Is this a bug?

```ruby
irb(main):001:0> Prism.parse('def foo(*); end').value.statements.body.first.parameters.rest
=> 
@ RestParameterNode (location: (1,8)-(1,9))                                                                                     
├── name: nil                                                                                                                   
├── name_loc: ∅                                                                                                                 
└── operator_loc: (1,8)-(1,9) = "*"

irb(main):002:0> Prism.parse('def foo((*)); end').value.statements.body.first.parameters.requireds.first.parameters.first
=> 
@ SplatNode (location: (1,9)-(1,10))                                                                                            
├── operator_loc: (1,9)-(1,10) = "*"                                                                                            
└── expression: ∅  
```

I would have expected a `RestParameterNode` in both cases, but maybe there's a semantic reason that a `SplatNode` has to be used in the second case?

**Comment by enebo** [comment]
In isolation it looks stranger but if you make a definition like:

```ruby
def foo((*), *)
end
```

then I think it looks more obvious it is just a splat in a destructuring of a required parameter.

**Comment by andrykonchin** [comment]
As for me `*` should lead to `RestParameterNode` in destructuring of a required parameter as well. At least for consistency.

`*` in parameters list in both cases is applied to a parameter optional name. Not local or instance variable, not arbitrary expression that are the options for `SplatNode`. But currently in case of destructured argument it applied to `RequiredParameterNode` (`(SplatNode (RequiredParameterNode))`) if parameter name is present and to `nil` otherwise (`(SplatNode nil)`).

If we want to reflect only syntax and ignore semantic - then any usage of `*` should be expressed with generic `SplatNode` and `RestParameterNode` seems excessive.

**Comment by enebo** [comment]
@andrykonchin After reading this I think at least one thing...What does Rest mean anymore?  It is an anachronistic term referring to Ruby 1.8 when no arguments could follow (e.g. it literally meant the rest vs some in the middle or the rest).  Post arguments at least make this term obsolete (although not for keyword rest where it still makes sense as a name).

I think my explanation also glossed over another point because I just put a code snippet and not the AST structure.  rest nodes are their own field in a structure.  We already know it is a rest node.  It could as easily be SplatNode as RestParmeterNode.  The other consideration is that we can get the name (or lack thereof) from it so we can build data for Method#parameters but we can also just get the name from Splat as easily as RestParameterNode.

I guess I am on board with removing RestParameter and just using Splat.  We should also consider whether rest is part of the lexicon forever or whether we should rename it in ParametersNode to be something like `splat`.

**Comment by eregon** [comment]
```
> method(def m(a, b=1, *c, d, e: 2, f:, **g, &h); end).parameters
=> [[:req, :a], [:opt, :b], [:rest, :c], [:req, :d], [:keyreq, :f], [:key, :e], [:keyrest, :g], [:block, :h]]
```

So I think the name `rest` is fine. It's still where the "rest of positional arguments" goes after the other parameters have been assigned (and similar for keyrest for keyword arguments).

**Comment by eregon** [comment]
Re the subject of this issue I'm not sure what is best.
I wonder if destructured parameters could always be treated as just extra statements in the body of the method.
So `def m((a,b,*), c); end` would become `def m(temp, c); a,b,* = temp; end`.
Of course there is the issue of temp variables, but maybe we could just consider some local variables with a given invalid prefix (e.g. `%`) to be temp variables.

I was also thinking it might not be OK if another parameter depends on a nested assignment but that seems fine:
```ruby
irb(main):004:0> def m((a,b), c=b); c; end
=> :m
irb(main):005:0> m([1,2])
=> nil
```

**Comment by kddnewton** [comment]
I agree for consistency's sake this is weird. I think we should make those `RestParameterNode` instances considering the other positions in the list are `RequiredParameterNode` instances.

Removing `RestParameterNode` can be considered separately, this is just a bug.

**Comment by eregon** [comment]
Maybe we should have a MultiWriteNode inside a RequiredDestructuredParameterNode?
I think that handling is very similar for `def m((a,*b,c)); end` and `a,*b,c = value`.
Also related https://github.com/ruby/prism/issues/1707 would make it easier to process such cases.
